Allow protocols to be nested in non-generic contexts

karwa opened the pull request:
<!-- What's in this pull request? -->
Allows protocols to be nested in non-generic types/functions.

Gated behind flag `-enable-experimental-feature NestedProtocols`. SE proposal coming soon.

TODO:

- [ ] Test that ASTPrinter adds feature guards around uses of nested protocols
- [ ] IRGen tests

hborla commented on lib/Sema/TypeCheckDeclPrimary.cpp:
```diff
@@ -2519,12 +2519,26 @@ class DeclChecker : public DeclVisitor<DeclChecker> {
+        // Protocols may only be nested in non-generic contexts.
+        if (NTD->getParent()->isGenericContext()) {
+          if (isa<ProtocolDecl>(NTD->getParent())) {
+            NTD->diagnose(diag::unsupported_nested_protocol_in_protocol, NTD->getName());
```
Do you also want this tailored diagnostic for the following cases?

```swift
protocol P {}

extension P {
  protocol Nested {}
}

extension P {
  func f() {
    protocol Nested {}
  }
}
```

karwa replied:
Yes, that's a good point. I added the tailored diagnostic because I don't think most Swift developers would think of a plain protocol as a "generic context".

karwa commented:
So, one thing that is interesting is the module interface printing - the logic in the PR is a bit too simplistic; if the goal is to print an interface which can be successfully parsed by an older compiler (using feature guards), then it's not enough to only guard the protocol declarations themselves -- it also needs to guard any _use_ of a nested protocol (e.g. in a function's parameters or return type, or within a generic signature). And I think that also extends to typealiases which are bound to nested protocols, across translation units, etc. So it's kind of complex.

What is the best way to approach that?

slavapestov commented:
> then it's not enough to only guard the protocol declarations themselves -- it also needs to guard any use of a nested protocol (e.g. in a function's parameters or return type, or within a generic signature).

Now imagine a declaration that transitively depends on a declaration that uses a nested protocol; it would also need to be feature-guarded, if you were going for an airtight implementation.

> What is the best way to approach that?

I would simply punt on the issue. Feature guards are only intended to solve a limited set of problems with staging in toolchain/stdlib/SDK changes and we can't really hope to always generate an interface file that remains parseable by older compilers.

slavapestov commented on test/decl/nested/protocol_experimental.swift:
```diff
@@ -0,0 +1,261 @@
+// RUN: %target-typecheck-verify-swift -parse-as-library -enable-experimental-feature NestedProtocols
```
Please also add some IRGen tests to cover emission of protocol descriptors, and execution tests. In particular, I'm curious if a dynamic cast to a conditional conformance involving one of these things will work immediately, or require runtime changes:

```swift
enum E {
  protocol P {}
}

struct S<T> {}
extension S: E.P where T: E.P {}
extension Int: E.P {}

let x: Any = S<Int>()
print(x as? any E.P)
```

karwa commented:
@slavapestov 

> I would simply punt on the issue. Feature guards are only intended to solve a limited set of problems with staging in toolchain/stdlib/SDK changes and we can't really hope to always generate an interface file that remains parseable by older compilers.

OK. I thought about maybe hiding just the protocol type, but the type declaration will almost always live alongside some uses of the protocol so it probably wouldn't be very useful in real projects, and in the worst case an older compiler could end up with an incorrect interface:

```swift
protocol Foo { ... }

enum MyEnum {
  #if compiler(>=5.3) && $NestedProtocols
    protocol Foo { ... }
  #endif

  func doSomething<T: Foo>(_: T)
  // ^ Argument may be top-level Foo or nested Foo depending on compiler version!
}
```

Probably (hopefully) that would end in a linker failure, but it's still not good. Maybe type declarations should not be guarded in general because of this. So yeah, in summary - I'll remove the feature guard in the module interface and add a test to ensure it isn't printed.

> In particular, I'm curious if a dynamic cast to a conditional conformance involving one of these things will work immediately, or require runtime changes

Your example works for me using this patch. I will add an executable test which includes it (as well as the IRGen descriptor tests).

slavapestov commented:
A couple more things I thought of:
- we probably want to ban @objc protocols from being nested
- we should test code generation and execution for protocols inside functions
- we should test binary serialization and deserialization

karwa commented:
> we probably want to ban @objc protocols from being nested

Why?

I mean, there is an issue that we don't namespace generated Obj-C names, but that is equally a problem for nested classes. For example, see the test `PrintAsObjC/classes`. It takes the Swift interface:

```swift
@objc @objcMembers class Nested {
  @objc @objcMembers class Inner {
    @objc @objcMembers class DeeperIn {}
  }
}
```

And generates the following Obj-C header. Note that the `SWIFT_CLASS` mangled names are correct (so that's a way to discriminate between `Foo.Inner` and `Bar.Inner`), but if somebody were to use the non-mangled name `Inner` in their Obj-C code, how would the Obj-C compiler know which class was intended?

```objc
SWIFT_CLASS("_TtC7classes6Nested")
@interface Nested
- (nonnull instancetype)init OBJC_DESIGNATED_INITIALIZER;
@end


SWIFT_CLASS("_TtCC7classes6Nested5Inner")
@interface Inner
- (nonnull instancetype)init OBJC_DESIGNATED_INITIALIZER;
@end


SWIFT_CLASS("_TtCCC7classes6Nested5Inner8DeeperIn")
@interface DeeperIn
- (nonnull instancetype)init OBJC_DESIGNATED_INITIALIZER;
@end
```

So clearly that's a problem, but it is not uniquely a problem for protocols and deserves a comprehensive solution. That solution would be source-breaking, and it's not going to be worse for protocols than any other nested type.

Evidently, some people seem to be able to work with the current limitations, and it's not clear to me that they should be prohibited from exposing nested protocols. They may wish to use a custom Obj-C name as a workaround, while using nesting for the Swift version of the interface.

slavapestov commented:
> I mean, there is an issue that we don't namespace generated Obj-C names, but that is equally a problem for nested classes. For example, see the test PrintAsObjC/classes. It takes the Swift interface:

Good point. Let's just add some test cases then.

karwa commented:
@slavapestov OK I've added:

- Refinements for the diagnostic
- Tests for PrintAsObjC (including a test for a nested classes with a custom Obj-C name)
- Test that IRGen emits protocol descriptors. I didn't actually base this on my old PR; I figured the compiler had moved on a lot and I wanted to rediscover what was needed so I had a fresh understanding, and I discovered (again) that if `parseDeclProtocol` doesn't set the local discriminator, we wouldn't emit protocol descriptors for function-local protocols. I checked what happens if I revert that change, and this test will catch it.

Still todo:

- ModuleInterface test (there should be no guarding of nested protocols)
- Serialisation/deserialisation test
- Executable test

And I think that's it, right?

slavapestov commented:
@swift-ci Please test

hborla commented:
@swift-ci please build toolchain

hborla commented:
@karwa I resolved conflicts via merge because I didn't want to force-push to your branch, but please feel free to redo it as a rebase if you'd like.

karwa committed:
[SE-0404] Allow protocols to be nested in non-generic contexts

slavapestov commented on lib/Sema/TypeCheckDeclPrimary.cpp:
```diff
@@ -3178,6 +3185,15 @@ class DeclChecker : public DeclVisitor<DeclChecker> {
     if (!PD->getPrimaryAssociatedTypeNames().empty())
       (void) PD->getPrimaryAssociatedTypes();
 
+    // We cannot compute the requirement signature of a protocol
```
Something else might trigger computation of the requirement signature though, even if we skip it here.

karwa replied:
The change you suggested to `getParentForLookup` solves this, so we don't need to skip it anymore.

slavapestov commented on lib/AST/DeclContext.cpp:
```diff
@@ -277,8 +281,8 @@ DeclContext *DeclContext::getInnermostSkippedFunctionContext() {
 DeclContext *DeclContext::getParentForLookup() const {
-  if (isa<ProtocolDecl>(this) || isa<ExtensionDecl>(this)) {
-    // If we are inside a protocol or an extension, skip directly
+  if (isa<ExtensionDecl>(this)) {
```
How about still returning getModuleScopeContext() here if we have an invalidly-nested protocol? That would fix your issue with superclass requirements.

slavapestov reviewed (commented):
I just have a couple more minor suggestions.

karwa committed:
[SE-0404] Rebase fixes and better handling of generic superclasses with inferred params

karwa commented:
@slavapestov Implemented the change you suggested. Could you trigger CI please?

slavapestov commented on test/decl/nested/protocol.swift:
```diff
@@ -115,15 +257,55 @@ func testLookup<T: OuterForUFI.Inner>(_ x: T) {
-  // expected-error@-1 {{protocol 'P' cannot be nested inside another declaration}}
+struct NestedInGenericMethod<T> {
+  func someMethod() {
+    protocol AnotherTest {} // expected-error{{type 'AnotherTest' cannot be nested in generic function 'someMethod()'}}
```
Shouldn't this use the "protocol %0 cannot be nested in a generic context" diagnostic?

karwa replied:
Another change (that I rebased in) guarded these checks inside `if (auto *parentDecl = DC->getSelfNominalTypeDecl())`. In the function, `getSelfNominalTypeDecl` returns null, so this goes along the regular nominal-in-generic-context error path. That path gives nicer diagnostics, so I left it that way.

slavapestov commented:
@swift-ci Please smoke test

karwa committed:
[SE-0404] Make test for protocol nested in actor conditionally-available

karwa commented:
Test failure is due to actor availability. Doesn't happen locally when I run the tests on Sonoma. I guess it happens because CI is using an ancient version of macOS?

```
Command Output (stderr):
--
/Users/ec2-user/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/swift/test/decl/nested/protocol.swift:52:13: error: unexpected error produced: concurrency is only available in macOS 10.15.0 or newer
      actor Level3 {
            ^
/Users/ec2-user/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/swift/test/decl/nested/protocol.swift:52:13: error: unexpected note produced: add @available attribute to enclosing actor
      actor Level3 {
            ^
```

Fixed by changing that actor be an enum. Added an additional test for nesting a protocol in an actor with an availability annotation.

slavapestov commented:
> I guess it happens because CI is using an ancient version of macOS?

I believe it's about the default deployment target in the SDK, but I forget the details.

> Fixed by changing that actor be an enum. Added an additional test for nesting a protocol in an actor with an availability annotation.

It is also fine to pass -disable-availability-checking in tests like this.

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
> /home/build-user/build/buildbot_linux/swift-linux-x86_64/test-linux-x86_64/Serialization/Output/nested-protocols.swift.script: line 3: Filecheck: command not found

Looks like this test fails on a case-sensitive file system.

karwa committed:
[Test] Fix use of 'FileCheck'

karwa commented:
@slavapestov Oops. But yeah, the macOS filesystem seems to be case-insensitive by default, so that didn't happen locally. At least that should be the last thing, since the macOS test was okay.

slavapestov commented:
@swift-ci Please smoke test

karwa commented:
OMG will Linux give me a break? Using substitution `%FileCheck` for FileCheck. 

Not needed on macOS because... 🤷‍♂️ who knows?

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@karwa Great job with the PR. I merged it since I don't think there is anything else outstanding here. Also, the proposal was accepted, so I think you can remove the experimental feature flag now.

slavapestov commented:
Also do you want to cherry-pick this to the release/5.10 branch?

karwa commented:
Thanks @slavapestov! And thanks for your guidance along the way!

Yeah I'll cherry-pick this to the 5.10 branch :)
